### ISSUE TITLE
Timing issue with initial admin code

### ISSUE THREAD

**Issue by matti** [kontena/kontena#1460]
When a new master is created there is a small time window that will invalidate the initial admin code for good. I think we've seen this on the slack channel as well?

repro (following the updated DEVELOPMENT.md) #1459 :

```
docker-compose up --service-ports api
```

another terminal (pseudobash):

```
while true; do
  timeout(1.5); do
    kontena master login --name devmaster --code initialadmincode http://localhost:9292
  end
end
```

This will result in:
```
 [error] invalid_grant
```

**Comment by matti** [comment]
and to clarify: grant will always be invalid, so you need to recreate the master.

**Comment by SpComb** [comment]
I've also seen this when deploying a new master, and having the initial `kontena master login` error for whatever reason - unable to re-login anymore.

Looks like the master invalidates the initial admin code in mongo after login.

#### `docker logs kontena-server-api`
```
I, [2016-12-01T09:28:56.638811 #6]  INFO -- Mongodb::Migrator: migrating CreateInitialAdmin
*********************************************
Initial Admin Code: rf...eq3
*********************************************
```

#### `docker exec -it kontena-server-mongo mongo kontena_development`
```
> db.access_tokens.find()
{ "_id" : ObjectId("583fed93b0783c000600000e"), "token_type" : "bearer", "internal" : true, "scopes" : [ "user", "owner" ], "code" : "rf...eq3", "user_id" : ObjectId("583fed93b0783c000600000d"), "expires_at" : null, "token" : "$2a$10$B6...dUS", "token_last_four" : "6350", "updated_at" : ISODate("2016-12-01T09:29:55.494Z"), "created_at" : ISODate("2016-12-01T09:29:55.494Z") }
```

The initial admin code is there

#### `kontena master login --code rf...eq3 http://.../`

#### `docker exec -it kontena-server-mongo mongo kontena_development`

```
> db.access_tokens.find()
{ "_id" : ObjectId("583fedb3b0783c0006000013"), "token_type" : "bearer", "internal" : true, "expires_at" : null, "scopes" : [ "user", "owner" ], "user_id" : ObjectId("583fed93b0783c000600000d"), "token" : "$2a$10$B6O...VDa", "token_last_four" : "617f", "updated_at" : ISODate("2016-12-01T09:30:27.407Z"), "created_at" : ISODate("2016-12-01T09:30:27.407Z") }
```

The intial admin code is gone

#### `DEBUG=true kontena master login --code rf...eq3 http://.../`
```
D, [2016-12-01T11:33:51.137361 #4158] DEBUG -- CONFIG: Loading configuration from /home/kontena/.kontena_client.json
D, [2016-12-01T11:33:51.137787 #4158] DEBUG -- CONFIG: Configuration loaded with 6 servers.
D, [2016-12-01T11:33:51.137816 #4158] DEBUG -- CONFIG: Current master: terom-xenial
D, [2016-12-01T11:33:51.137851 #4158] DEBUG -- CONFIG: Current grid: terom-xenial
Running #<Kontena::MainCommand:0x00000001606ae8> -- callback matcher = 'nil'
Running #<Kontena::Cli::MasterCommand:0x000000015fdc90> -- callback matcher = 'nil'
Running #<Kontena::Cli::Master::LoginCommand:0x000000015e8840> -- callback matcher = 'master login'
D, [2016-12-01T11:33:51.230786 #4158] DEBUG -- CLIENT: Excon opts: {:omit_default_port=>true, :connect_timeout=>5, :read_timeout=>30, :write_timeout=>5, :ssl_verify_peer=>true}
D, [2016-12-01T11:33:51.346345 #4158] DEBUG -- CLIENT: Request POST /oauth2/token: 404 Not Found: {"error":"invalid_grant","error_description":null}
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/client.rb:512:in `handle_error_response': invalid_grant (Kontena::Errors::StandardError)
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/client.rb:342:in `rescue in request'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/client.rb:283:in `request'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/client.rb:148:in `exchange_code'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/cli/master/login_command.rb:104:in `block in use_authorization_code'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/cli/common.rb:80:in `vspinner'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/cli/master/login_command.rb:103:in `use_authorization_code'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/cli/master/login_command.rb:49:in `execute'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/command.rb:187:in `run'
	from /home/kontena/.gem/ruby/2.3.0/gems/clamp-1.1.1/lib/clamp/subcommand/execution.rb:11:in `execute'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/command.rb:187:in `run'
	from /home/kontena/.gem/ruby/2.3.0/gems/clamp-1.1.1/lib/clamp/subcommand/execution.rb:11:in `execute'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/lib/kontena/command.rb:187:in `run'
	from /home/kontena/.gem/ruby/2.3.0/gems/clamp-1.1.1/lib/clamp/command.rb:132:in `run'
	from /home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.0.0/bin/kontena:16:in `<top (required)>'
	from /home/kontena/.local/bin/kontena:23:in `load'
	from /home/kontena/.local/bin/kontena:23:in `<main>'
```

**Comment by kke** [comment]
@SpComb yes that's how it's supposed to work, it's a one time oauth2 authorization_code.

I think the problem here is that the API starts answering before all the migrations have finished.

**Comment by jakolehm** [comment]
@kke server is running migrations in the `on_worker_boot` puma hook. It should wait until migrations are completed before accepting any connections.

**Comment by SpComb** [comment]
@jakolehm was this closed by #1921? Something very similar with `invalid_grant` errors on `kontena master login --code` is still happening in 1.2: #2295
